presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestDistributedQueries.java

```diff
@@ -1253,7 +1257,7 @@ public void testPayloadJoins()
                 "SELECT l.* FROM (select orderkey+1 as ok1, * from lineitem_map) l left join orders o on (l.ok1 = o.orderkey+1) left join part p on (l.partkey=p.partkey)",
                 "SELECT l.* FROM (select *, cast(orderkey as int) as ok from lineitem_map) l left join (select *, cast(orderkey as int) as ok from orders) o on (l.ok = o.ok) left join (select *, cast(partkey as int) as pk from part) p on (l.ok=p.pk)",
                 "with lm as (select quantity + 1 as q1, * from lineitem_map) SELECT l.* FROM (select * from lm) l left join orders o on (l.orderkey = o.orderkey) left join part p on (l.partkey=p.partkey)",
-
+                // This is the query which fail when REMOVE_REDUNDANT_CAST_TO_VARCHAR_IN_JOIN set to true
                 "SELECT l.* FROM lineitem_map l left join orders o on (cast(l.orderkey as varchar) = cast(o.orderkey as varchar)) left join part p on (l.partkey=p.partkey)",
```

Comment:
@mlyublena The optimization added surfaced a bug in payload join optimizer, i.e. when payload join optimization is not enabled, this query runs successfully with the optimization in this PR, but when payload join optimization is enabled, this query will fail with the optimization in this PR.

Reply:
There is an outstanding PR that fixes a lot of bugs in payload join optimizer: https://github.com/prestodb/presto/pull/19904 
My guess is that the case you found is already fixes there